#!/usr/bin/env Rscript

#
# Created by Pierluigi Di Chiaro
#

## This tool proceeded with the creation of a reference database


library("optparse")

option_list = list(
  make_option(c("-i", "--master_file"), type="character", default=NULL,
              help="provide the master file with all info",  metavar="character"),
  make_option(c("-o", "--output"), type="character", default=NULL,
              help="provide output folder", metavar="character"),
  make_option(c("-u", "--usefuldata"), type="character", default=NULL,
              help="specify the UsefulData directory where DB files(genome, GTF...) are stored", metavar="character")
)

optparser = OptionParser(option_list=option_list);
opt = parse_args(optparser)

if (is.null(opt$master)) {print_help(optparser)
  stop("At least one argument must be supplied (master_file)", call.=FALSE)}
if (is.null(opt$output)) {print_help(optparser)
  stop("At least one argument must be supplied (output folder)", call.=FALSE)}
if (is.null(opt$usefuldata)) {print_help(optparser)
  stop("At least one argument must be supplied (usefuldata folder)", call.=FALSE)}


library("tidyverse")
library("tximport")
library("knitr")
library("rmarkdown")
library("GenomeInfoDb")
library("Rsamtools")
library("GenomicAlignments")
library("BiocParallel")
library("Rsubread")
library("GenomicFeatures")
library("rtracklayer")


MASTER_FILE <- read.delim(opt$master, header=TRUE)
Output_folder <- opt$output                                                   
UsefulData <- opt$usefuldata

Norm_folder <- paste0(Output_folder, "/6_Norm_folder/")
Counts_folder  <- paste0(Output_folder, "/7_Counts_folder/")

dir.create(Counts_folder, showWarnings = FALSE)
dir.create(Norm_folder, showWarnings = FALSE)

reference_folder <- paste0(UsefulData, "/DB/hg38_annotation/")
gencode_folder <- paste0(UsefulData, "/DB/iGenome/Homo_sapiens/UCSC/hg38/Annotation/hg38_gencode_annotation/")
BAM_DIR <- paste0(Output_folder,"/4_Bam")


## ---- Associate transcripts to gene IDs by tximport ---------
## Consider abbundance.h5 generated by Kallisto 
cat("Associating transcripts to gene IDs\n")

gencode_df <- as.data.frame(rtracklayer::import(paste0(gencode_folder,"gencode.v33.basic.annotation.gtf")))
gencode_transcript <- dplyr::filter(gencode_df,type=="transcript")
gencode_gene <- dplyr::filter(gencode_df,type=="gene")
tx2gene <- dplyr::select(gencode_transcript, transcript_id, gene_id)

samples <- dir(file.path(BAM_DIR))
path <- file.path(BAM_DIR, samples, 'abundance.h5')

txi.kallisto <- tximport(path, type = "kallisto", tx2gene = tx2gene)
all.reads <- as.data.frame(txi.kallisto$counts)
names(all.reads) <- samples

## ---- Create the reference on which we will record the Tx changes ---------
cat("Create a transcript database\n")

GENES <- as.data.frame(matrix(NA,nrow=nrow(all.reads)))
rownames(GENES) <- rownames(all.reads)
colnames(GENES) <- "gene_id"

GENES$gene_id <- as.character(rownames(GENES))
GENES$gene_ids <- gsub(GENES$gene_id, pattern="\\.[0-9]+$|\\.[0-9]+_PAR_Y", replacement="")
GENES$IDs	<- NA
GENES$tx_id <- NA
GENES$seqnames <- NA
GENES$start <- NA
GENES$end <- NA
GENES$width <- NA
GENES$strand <- NA
GENES$LOC	<- NA
GENES$type_of_gene <- NA
GENES$Symbol <- NA
GENES$HGNC <- NA
GENES$Name <- NA
GENES$Entrez.Gene.ID <- NA
GENES$cDNA_LENGTH <- NA

for(ii in seq_along(GENES$gene_id)){
  cat(ii," of ",length(GENES$gene_id),"\n")
  ww <- which(gencode_gene$gene_id == GENES$gene_id[ii])
  if(length(ww)!=0){
    cat("\t\t Found \n")
    GENES$IDs[ii]	<- as.character(gencode_gene$gene_name[ww])
  }
}

dup <- GENES$IDs[duplicated(GENES$gene_ids)]

for(ii in dup){
  w <- which(GENES$IDs==ii)
    if(length(w) == 1){
      cat(ii," not duplicated \n")
    }else{
      cat(ii," is duplicated \n")
        GENES[w[1],"IDs"] <- paste0(ii,":1")
        GENES[w[2],"IDs"] <- paste0(ii,":2")
    }
  }

## Import the reference file - reporting symbol, Name, location etc.. of NCBI, HGNC and ENSEMBL database
Ref_files <- list.files(reference_folder,full.names = TRUE, recursive = FALSE,ignore.case = FALSE, include.dirs = TRUE)
Anno_Txdb <- read.delim(Ref_files[grep("_merged_annotation.txt",Ref_files)])  # Add annotations Anno_Txdb
rownames(Anno_Txdb) <- Anno_Txdb$X

for(ii in seq_along(GENES$gene_ids)){
  cat(ii," of ",length(GENES$gene_ids),"\n")
  ww <- grep(GENES$gene_ids[ii],gsub("\\;ENSG*", "", Anno_Txdb$ENSEMBL))
  if(length(ww)==1){
    cat("\t\t Found \n")
    GENES$tx_id[ii] <- as.character(Anno_Txdb$tx_id[ww])
    GENES$seqnames[ii] <- as.character(Anno_Txdb$seqnames[ww])
    GENES$start[ii] <- as.character(Anno_Txdb$start[ww])
    GENES$end[ii] <- as.character(Anno_Txdb$end[ww])
    GENES$width[ii] <- as.character(Anno_Txdb$width[ww])
    GENES$strand[ii] <- as.character(Anno_Txdb$strand[ww])
    GENES$LOC[ii]	<- as.character(Anno_Txdb$LOC[ww])
    GENES$type_of_gene[ii] <- as.character(Anno_Txdb$type_of_gene[ww])
    GENES$Symbol[ii] <- as.character(Anno_Txdb$Symbol[ww])
    GENES$HGNC[ii] <- as.character(Anno_Txdb$HGNC[ww])
    GENES$Name[ii] <- as.character(Anno_Txdb$Name[ww])
    GENES$Entrez.Gene.ID[ii] <- as.character(Anno_Txdb$Entrez.Gene.ID[ww])
    GENES$cDNA_LENGTH[ii] <- as.character(Anno_Txdb$cDNA_LENGTH[ww])
  }else{
    ww <- grep(GENES$gene_ids[ii],gsub("\\;ENSG*", "", Anno_Txdb$ENSEMBL_CHECK))
    if(length(ww)==1){
      cat("\t\t Found \n")
      GENES$tx_id[ii] <- as.character(Anno_Txdb$tx_id[ww])
      GENES$seqnames[ii] <- as.character(Anno_Txdb$seqnames[ww])
      GENES$start[ii] <- as.character(Anno_Txdb$start[ww])
      GENES$end[ii] <- as.character(Anno_Txdb$end[ww])
      GENES$width[ii] <- as.character(Anno_Txdb$width[ww])
      GENES$strand[ii] <- as.character(Anno_Txdb$strand[ww])
      GENES$LOC[ii]	<- as.character(Anno_Txdb$LOC[ww])
      GENES$type_of_gene[ii] <- as.character(Anno_Txdb$type_of_gene[ww])
      GENES$Symbol[ii] <- as.character(Anno_Txdb$Symbol[ww])
      GENES$HGNC[ii] <- as.character(Anno_Txdb$HGNC[ww])
      GENES$Name[ii] <- as.character(Anno_Txdb$Name[ww])
      GENES$Entrez.Gene.ID[ii] <- as.character(Anno_Txdb$Entrez.Gene.ID[ww])
      GENES$cDNA_LENGTH[ii] <- as.character(Anno_Txdb$cDNA_LENGTH[ww])
    }else{
      if(length(ww)!=0){
        ww <- which(rownames(Anno_Txdb) == GENES$IDs[ii])
        if(length(ww)==1){
          cat("\t\t Found \n")
          GENES$tx_id[ii] <- as.character(Anno_Txdb$tx_id[ww])
          GENES$seqnames[ii] <- as.character(Anno_Txdb$seqnames[ww])
          GENES$start[ii] <- as.character(Anno_Txdb$start[ww])
          GENES$end[ii] <- as.character(Anno_Txdb$end[ww])
          GENES$width[ii] <- as.character(Anno_Txdb$width[ww])
          GENES$strand[ii] <- as.character(Anno_Txdb$strand[ww])
          GENES$LOC[ii]	<- as.character(Anno_Txdb$LOC[ww])
          GENES$type_of_gene[ii] <- as.character(Anno_Txdb$type_of_gene[ww])
          GENES$Symbol[ii] <- as.character(Anno_Txdb$Symbol[ww])
          GENES$HGNC[ii] <- as.character(Anno_Txdb$HGNC[ww])
          GENES$Name[ii] <- as.character(Anno_Txdb$Name[ww])
          GENES$Entrez.Gene.ID[ii] <- as.character(Anno_Txdb$Entrez.Gene.ID[ww])
          GENES$cDNA_LENGTH[ii] <- as.character(Anno_Txdb$cDNA_LENGTH[ww])
        }
      }
    }
  }
}

set <- which(is.na(GENES$Symbol))
if(length(set)!=0){
  GENES <- GENES[-set,]
}

## Filter out genes on the chrY PAR regions
set <- grep("_PAR_Y",GENES$gene_id)
if(length(set)!=0){
  GENES <- GENES[-set,]
}

## Mantain only genes which are present in NCBI and ENSEMBL
set <- which(GENES$Symbol == gsub("\\:[0-9]+$","",GENES$IDs))
if(length(set)!=0){
  GENES <- GENES[set,]
}


## ---- Filter exonic reads count generated by Kallisto using the new reference ---------
cat("Associating reference to gene counts\n")

reads <- all.reads[rownames(GENES),samples]
RES_DAT_ALL <- cbind(GENES, reads)
RES_DAT_ALL <- RES_DAT_ALL[order(as.numeric(RES_DAT_ALL$tx_id)),]
RES_DAT_ALL$tx_id <- NULL
rownames(RES_DAT_ALL) <- NULL

write.table(RES_DAT_ALL,file=paste0(Counts_folder,"EX_reads_RAW.txt"),sep="\t",col.names=NA)

